mypy/checkexpr.py

```diff
@@ -3411,6 +3411,10 @@ def visit_yield_from_expr(self, e: YieldFromExpr, allow_none_return: bool = Fals
         elif self.chk.type_is_iterable(subexpr_type):
             if is_async_def(subexpr_type) and not has_coroutine_decorator(return_type):
                 self.chk.msg.yield_from_invalid_operand_type(subexpr_type, e)
+            elif isinstance(subexpr_type, TupleType):
+                # Tuples get special handling here, since they "don't have __iter__"...
```

Comment:
This is not true, tuples have an `__iter__` (on the fallback).

Reply:
Do they? I haven't been able to find it:

```py
>>> subexpr_type
Tuple[]
>>> tuple_fallback(subexpr_type)
builtins.tuple[<nothing>]
>>> tuple_fallback(subexpr_type).type
<TypeInfo builtins.tuple>
>>> tuple_fallback(subexpr_type).type.mro
[<TypeInfo builtins.tuple>, <TypeInfo builtins.object>]
>>> tuple_fallback(subexpr_type).type.mro[0].names
{}
>>> tuple_fallback(subexpr_type).type.mro[1].names
{'__init__': <mypy.nodes.SymbolTableNode object at 0x1137ce5f8>}
```

The same is true of both full and empty tuples.

Reply:
> Do they?

Yes, they do. Please use a reasonable/relevant test fixture instead of the default one, e.g. `fixtures/tuple.pyi`.

Reply:
Thanks @ilevkivskyi! I hadn’t realized that was why it was behaving so strangely. It looks like initial attempts to fix it in the issue (#4444) made the same mistake, which was what led me down this path.

You’re right that this branching is no longer needed then. Fixing `type_is_iterable` solves the issue!